### PR TITLE
feat: proxy mode support for BaseStrategy via BaseStrategyInitializable

### PR DESCRIPTION
**Pull request by emilianobonassi**
Reduce deployment costs if you have template strategies. Instead of inherit, delegate storage for params to minimal proxies.

e.g. https://github.com/emilianobonassi/yIdleStrategies/blob/dev/contracts/StrategyIdle.sol

### CODE REVIEW

**Comment by fubuloubu** [comment]
Instead of making this modification, can you add another contract `BaseStrategyInitializable` that is defined in the currenct `BaseStrategy.sol`, extending `BaseStrategy` and just adds an `initialize` function which does the same thing as the `constructor`?

This way, you just inherit from `BaseStrategyInitializable` (instead of `BaseStrategy`) if you want this feature

**Review by fubuloubu — CHANGES_REQUESTED** [review]
Looks good but needs a test

**Comment by emilianobonassi** [comment]
Thanks @fubuloubu 

Test for sure

I will add a blocker to call `init` twice (a simple book) with the respective test as I have in the idle strats

File: contracts/BaseStrategy.sol [hunk 1]
```diff
@@ -236,21 +236,30 @@ abstract contract BaseStrategy {
...
+        address _keeper
+    ) internal {
         vault = VaultAPI(_vault);
         want = IERC20(vault.token());
         want.approve(_vault, uint256(-1)); // Give Vault unlimited access (might save gas)
-        strategist = msg.sender;
-        rewards = msg.sender;
-        keeper = msg.sender;
-        vault.approve(rewards, uint256(-1)); // Allow rewards to be pulled
+        strategist = _strategist;
```
Comment (fubuloubu) [inline contracts/BaseStrategy.sol:259]: Add `require(address(vault) == 0x0);` here

**Comment by emilianobonassi** [comment]
TODO:
- test basestrategy via minimal proxy works as intended (use similar changes did on tests for on erc20 non compliant)
- test cannot be re-initialized

File: tests/functional/strategy/test_config.py [hunk 2]
```diff
@@ -33,6 +33,13 @@ def test_strategy_deployment(strategist, vault, TestStrategy):
     assert not strategy.tendTrigger(0)
 
 
+def test_strategy_no_reinit(strategist, vault, TestStrategy):
+    strategy = strategist.deploy(TestStrategy, vault)
+
+    with brownie.reverts("Strategy already initialized"):
+        strategy.initialize(vault, strategist, strategist, strategist)
+
```
Comment (fubuloubu) [inline tests/functional/strategy/test_config.py:41]: add a test here deploying a new proxy-version from the already deployed version, and show re-init doesn't work there either.

Comment (emilianobonassi) [reply tests/functional/strategy/test_config.py:41]: I was preparing it, let's see the next commit

**Comment by emilianobonassi** [comment]
Done, just looking for green light from CI 🚀

**Review by fubuloubu — APPROVED** [review]

**Review by fubuloubu — APPROVED** [review]

**Pull request merged**